pricegraph/src/orderbook.rs

```diff
@@ -140,6 +141,69 @@ impl Orderbook {
         debug_assert!(!self.is_overlapping());
     }
 
+    /// Reduces all overlapping orders for a given market and returns the
+    /// computed overlapping transitive orderbook for that market.
+    pub fn reduce_overlapping_transitive_orderbook(
+        &mut self,
+        base: TokenId,
+        quote: TokenId,
+    ) -> TransitiveOrderbook {
+        let (base, quote) = (node_index(base), node_index(quote));
+
+        let mut overlap = TransitiveOrderbook::default();
+        while let Err(NegativeCycle(predecessors, node)) =
+            bellman_ford::search(&self.projection, quote)
+        {
+            let path = {
+                let mut cycle = path::find_cycle(&predecessors, node, Some(quote))
+                    .expect("negative cycle not found after being detected");
+                cycle.push(cycle[0]);
+                cycle
+            };
+
+            if path[0] == quote {
+                if let Some(base_index) = path.iter().position(|node| *node == base) {
```

Comment:
nit: could maybe use `path.split` here?

Reply:
Unfortunately, we need slightly different semantics, as we want the `base` token to be in both sub-paths.